File: go/tokens/tokens.go
// Package tokens mints NotifyService inbox/prefs delegated JWTs. Server-side only.
package tokens

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"errors"
	"strings"
	"time"
)

const (
	DefaultInboxTTL = 15 * time.Minute
	DefaultPrefsTTL = 900 * time.Second
	minTTL          = 60 * time.Second
	maxTTL          = 24 * time.Hour
)

// Options are the inputs for minting a delegated token.
type Options struct {
	OrganizationID string
	ExternalUserID string
	SigningKey      string
	TTL            time.Duration // zero => scope default
}

// Payload is the decoded JWT body.
type Payload struct {
	Iss   string `json:"iss"`
	Sub   string `json:"sub"`
	Scope string `json:"scope"`
	Iat   int64  `json:"iat"`
	Exp   int64  `json:"exp"`
}

// MintInbox mints a token with scope "inbox".
func MintInbox(o Options) (string, error) { return mint(o, "inbox", DefaultInboxTTL) }

// MintPrefs mints a token with scope "prefs".
func MintPrefs(o Options) (string, error) { return mint(o, "prefs", DefaultPrefsTTL) }

// MintInboxAndPrefs mints a token with scope "inbox prefs".
func MintInboxAndPrefs(o Options) (string, error) { return mint(o, "inbox prefs", DefaultPrefsTTL) }

// Decode parses the payload without verifying the signature. Debugging only.
func Decode(token string) (Payload, error) {
	parts := strings.Split(token, ".")
	if len(parts) != 3 {
		return Payload{}, errors.New("token must have three dot-separated parts")
	}
	raw, err := base64.RawURLEncoding.DecodeString(parts[1])
	if err != nil {
		return Payload{}, err
	}
	var p Payload
	if err := json.Unmarshal(raw, &p); err != nil {
		return Payload{}, err
	}
	return p, nil
}

func mint(o Options, scope string, defaultTTL time.Duration) (string, error) {
	ttl := o.TTL
	if ttl == 0 {
		ttl = defaultTTL
	}
	if ttl < minTTL || ttl > maxTTL {
		return "", errors.New("ttl must be between 60s and 24h")
	}
	if !strings.HasPrefix(o.SigningKey, "nsi_") {
		return "", errors.New("signing key must start with 'nsi_'")
	}
	now := time.Now().Unix()
	payload, err := json.Marshal(Payload{Iss: o.OrganizationID, Sub: o.ExternalUserID, Scope: scope, Iat: now, Exp: now + int64(ttl.Seconds())})
	if err != nil {
		return "", err
	}
	headerB64 := base64.RawURLEncoding.EncodeToString([]byte(`{"alg":"HS256","typ":"JWT"}`))
	payloadB64 := base64.RawURLEncoding.EncodeToString(payload)
	signingInput := headerB64 + "." + payloadB64
	mac := hmac.New(sha256.New, []byte(o.SigningKey))
	mac.Write([]byte(signingInput))
	return signingInput + "." + base64.RawURLEncoding.EncodeToString(mac.Sum(nil)), nil
}

File: python/notifyservice/tokens.py
"""Delegated inbox and preferences token minting for NotifyService."""
import base64
import hashlib
import hmac
import json
import time
from dataclasses import dataclass

_DEFAULT_INBOX_TTL = 900
_DEFAULT_PREFS_TTL = 900
_MIN_TTL = 60
_MAX_TTL = 24 * 60 * 60


@dataclass(frozen=True)
class TokenPayload:
    iss: str
    sub: str
    scope: str
    iat: int
    exp: int


def mint_inbox_token(organization_id, external_user_id, signing_key, ttl_seconds=_DEFAULT_INBOX_TTL):
    return _mint(organization_id, external_user_id, signing_key, "inbox", ttl_seconds)


def mint_prefs_token(organization_id, external_user_id, signing_key, ttl_seconds=_DEFAULT_PREFS_TTL):
    return _mint(organization_id, external_user_id, signing_key, "prefs", ttl_seconds)


def mint_inbox_and_prefs_token(organization_id, external_user_id, signing_key, ttl_seconds=_DEFAULT_PREFS_TTL):
    return _mint(organization_id, external_user_id, signing_key, "inbox prefs", ttl_seconds)


def decode(token: str) -> TokenPayload:
    parts = token.split(".")
    if len(parts) != 3:
        raise ValueError("Token must have three dot-separated parts.")
    payload = json.loads(_b64url_decode(parts[1]))
    return TokenPayload(payload["iss"], payload["sub"], payload["scope"], payload["iat"], payload["exp"])


def _mint(organization_id, external_user_id, signing_key, scope, ttl_seconds):
    if ttl_seconds < _MIN_TTL or ttl_seconds > _MAX_TTL:
        raise ValueError(f"ttl_seconds must be between {_MIN_TTL} and {_MAX_TTL}.")
    if not signing_key.startswith("nsi_"):
        raise ValueError("signing_key must start with 'nsi_'.")

    now = int(time.time())
    header = {"alg": "HS256", "typ": "JWT"}
    payload = {"iss": str(organization_id), "sub": external_user_id, "scope": scope, "iat": now, "exp": now + ttl_seconds}

    header_b64 = _b64url(json.dumps(header, separators=(",", ":")).encode("utf-8"))
    payload_b64 = _b64url(json.dumps(payload, separators=(",", ":")).encode("utf-8"))
    signing_input = f"{header_b64}.{payload_b64}"
    sig = hmac.new(signing_key.encode("utf-8"), signing_input.encode("utf-8"), hashlib.sha256).digest()
    return f"{signing_input}.{_b64url(sig)}"


def _b64url(data: bytes) -> str:
    return base64.urlsafe_b64encode(data).rstrip(b"=").decode("ascii")


def _b64url_decode(s: str) -> bytes:
    return base64.urlsafe_b64decode(s + "=" * ((-len(s)) % 4))
